### PR TITLE
Fix unsanitized output of scan with strings

### LINKED ISSUE #13449: [BUG] cudf::scan on a String column returns a Column with non-empty nulls
**Describe the bug**
Creating a String column without non-empty nulls and running a `max` aggregation using scan on it returns a column_view with non-empty nulls

**Steps/Code to reproduce bug**
```
  @Test
  public void testScanHasEmptyNulls() {
    try (ColumnVector strings = ColumnVector.fromStrings("ãçÂL­", "âæb÷(´û", "Ôq=`È§:",
            null, "This", "is", "a", "String", null, "column", ".");
         ColumnVector vec = strings.scan(ScanAggregation.max(), ScanType.INCLUSIVE, NullPolicy.EXCLUDE)) {
      assertFalse(strings.hasNonEmptyNulls());
      assertFalse(vec.hasNonEmptyNulls()); // failing assertion
    }
  }
```

**Expected behavior**
The above test should pass

**Environment overview (please complete the following information)**
 - Environment location: Bare-metal
 - Method of cuDF install: source

**Environment details**
<details><summary>Click here to see environment details</summary>
<pre>
     **git***
     commit c601b83f3e03b3a1ef489daafb0dd520b3fdbe29 (HEAD)
     Author: Alessandro Bellina <abellina@nvidia.com>
     Date:   Thu May 18 13:17:59 2023 -0500
     
     Implement a chunked_pack API (#13260)
     
     This PR introduces a "chunked pack" API built on top of `contiguous_split`. This API is used when we want to copy a cuDF `table_view` over the wire or to the host in a contiguous layout (aka `contiguous_split`), but without user provided memory. As a result this API does not allocate any buffers for GPU data and instead it uses a provided user buffer to perform the `contiguous_split` in chunks.
     
     Luckily, `contiguous_split` already had a subdivision of work that we are now calling "batches". Each batch is up to 1MB of data from the source table. As such, one can think of this function as copying as many batches as will fit in a user buffer (or "chunking the batches"). The API follows other chunked interfaces in cuDF with a `has_next` and a `next`, with the difference that in this case `next` takes a `device_span`, and the user can provide any `device_span` as long as the size is the same as the size specified during construction.
     
     When thrust and scratch space is required on the GPU, this PR makes use of the memory resource passed, even to the point of using the second argument of `exec_policy` which is the memory resource. I found this while testing that `exec_policy` defaults to the per-device resource, and in this case I really wanted to pass a pooled memory resource, outside of our normal async memory resource to set aside this memory ahead of time.
     
     Most of the changes are about moving things around. You'll see 3 structs with the name "packed" (e.g. `packed_split_indices_and_src_buf_info`). These are here to group together state that `contiguous_split` needs to work, but now because of `chunked_pack` we need to keep around as well for the subsequent calls to `next`. These structs are also packed in memory, which is an optimization contiguous_split had already done to reduce the number of d2h/h2d copies. This PR did need to add a state object (`contiguous_split_state`) that now `contiguous_split` leverages. It also makes use of the `metadata_builder` which we added in a prior PR.
     
     This PR does not include the JNI changes needed for this to work on the java side, I'll post that separately. I figured this was too big already (and if people have suggestions on "chunking" this PR up, I am happy to do that).
     
     @nvdbaranec spent a great deal of time documenting `contiguous_split` for me and he suggested a path to get this done that I just followed (thank you!!)
     
     Authors:
     - Alessandro Bellina (https://github.com/abellina)
     
     Approvers:
     - https://github.com/nvdbaranec
     - Nghia Truong (https://github.com/ttnghia)
     - David Wendt (https://github.com/davidwendt)
     - Vyas Ramasubramani (https://github.com/vyasr)
     
     URL: https://github.com/rapidsai/cudf/pull/13260
     **git submodules***
     
     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=22.04
     DISTRIB_CODENAME=jammy
     DISTRIB_DESCRIPTION="Ubuntu 22.04.2 LTS"
     PRETTY_NAME="Ubuntu 22.04.2 LTS"
     NAME="Ubuntu"
     VERSION_ID="22.04"
     VERSION="22.04.2 LTS (Jammy Jellyfish)"
     VERSION_CODENAME=jammy
     ID=ubuntu
     ID_LIKE=debian
     HOME_URL="https://www.ubuntu.com/"
     SUPPORT_URL="https://help.ubuntu.com/"
     BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
     PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
     UBUNTU_CODENAME=jammy
     Linux raza-linux-2 5.15.0-72-generic #79-Ubuntu SMP Wed Apr 19 08:22:18 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
     
     ***GPU Information***
     Thu May 25 09:42:23 2023
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |                               |                      |               MIG M. |
     |===============================+======================+======================|
     |   0  NVIDIA TITAN X ...  Off  | 00000000:05:00.0  On |                  N/A |
     | 29%   53C    P0    59W / 250W |    868MiB / 12288MiB |      1%      Default |
     |                               |                      |                  N/A |
     +-------------------------------+----------------------+----------------------+
     |   1  Quadro RTX 6000     Off  | 00000000:0A:00.0 Off |                  Off |
     | 33%   38C    P8    28W / 260W |      6MiB / 24576MiB |      0%      Default |
     |                               |                      |                  N/A |
     +-------------------------------+----------------------+----------------------+
     
     +-----------------------------------------------------------------------------+
     | Processes:                                                                  |
     |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
     |        ID   ID                                                   Usage      |
     |=============================================================================|
     |    0   N/A  N/A      2564      G   /usr/lib/xorg/Xorg                388MiB |
     |    0   N/A  N/A      2738      G   /usr/bin/gnome-shell              146MiB |
     |    0   N/A  N/A      3477      G   ...RendererForSitePerProcess        2MiB |
     |    0   N/A  N/A      5479      G   ...7/usr/lib/firefox/firefox      225MiB |
     |    0   N/A  N/A     13268      G   ...veSuggestionsOnlyOnDemand       79MiB |
     |    1   N/A  N/A      2564      G   /usr/lib/xorg/Xorg                  4MiB |
     +-----------------------------------------------------------------------------+
     
     ***CPU***
     Architecture:                    x86_64
     CPU op-mode(s):                  32-bit, 64-bit
     Address sizes:                   46 bits physical, 48 bits virtual
     Byte Order:                      Little Endian
     CPU(s):                          12
     On-line CPU(s) list:             0-11
     Vendor ID:                       GenuineIntel
     Model name:                      Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz
     CPU family:                      6
     Model:                           63
     Thread(s) per core:              2
     Core(s) per socket:              6
     Socket(s):                       1
     Stepping:                        2
     CPU max MHz:                     3600.0000
     CPU min MHz:                     1200.0000
     BogoMIPS:                        6598.47
     Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d
     Virtualization:                  VT-x
     L1d cache:                       192 KiB (6 instances)
     L1i cache:                       192 KiB (6 instances)
     L2 cache:                        1.5 MiB (6 instances)
     L3 cache:                        15 MiB (1 instance)
     NUMA node(s):                    1
     NUMA node0 CPU(s):               0-11
     Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
     Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
     Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
     Vulnerability Meltdown:          Mitigation; PTI
     Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
     Vulnerability Retbleed:          Not affected
     Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
     Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
     Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
     Vulnerability Srbds:             Not affected
     Vulnerability Tsx async abort:   Not affected
     
     ***CMake***
     /usr/local/bin/cmake
     cmake version 3.20.2
     
     CMake suite maintained and supported by Kitware (kitware.com/cmake).
     
     ***g++***
     /usr/local/bin/g++
     g++ (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
     Copyright (C) 2021 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     
     
     ***nvcc***
     /usr/local/bin/nvcc
ccache: error: Could not find compiler "nvcc" in PATH
     
     ***Python***
     
     ***Environment Variables***
     PATH                            : /opt/maven/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin:/usr/local/cuda/bin
     LD_LIBRARY_PATH                 :
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    :
     PYTHON_PATH                     :
     
     conda not found
     ***pip packages***
     /usr/local/bin/pip
./print_env.sh: /usr/local/bin/pip: /usr/bin/python: bad interpreter: No such file or directory

```

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes output from `cudf::scan` to return sanitized null entries. The null rows are expected to contain no characters if the input null rows are also empty.
The code logic is changed to compute the output validity mask first and use the mask to help build the gather map used for creating the output strings column. This rework also cleans up an unused parameter from internal functions `detail::scan_inclusive` and `detail::scan_exclusive`.
An additional check is added to the `scan_test` utility to verify the results contain empty null entries.

Closes #13449 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by GregoryKimball** [comment]
Thank you @davidwendt!! That was fast!

**Review by PointKernel — APPROVED** [review]

File: cpp/src/reductions/scan/scan_inclusive.cu [hunk 1]
```diff
@@ -142,10 +149,23 @@ struct scan_functor<Op, cudf::string_view> {
...
+      // fill the null entries with out-of-bounds value so gather records them as null
+      // this prevents un-sanitized null entries in the output
+      auto null_itr = detail::make_counting_transform_iterator(0, null_iterator{mask});
+      auto oob_val  = thrust::constant_iterator<size_type>(input_view.size());
+      thrust::scatter_if(rmm::exec_policy(stream),
+                         oob_val,
+                         oob_val + input_view.size(),
+                         thrust::counting_iterator<size_type>(0),
+                         null_itr,
+                         result.data());
```
Comment (ttnghia) [inline cpp/src/reductions/scan/scan_inclusive.cu:162]: `result` here is not the function result. It is just an intermediate gather map. So we should better rename it into `result_map` or `gather_map` or something similar.

**Review by ttnghia — APPROVED** [review]

**Comment by davidwendt** [comment]
@razajafri Do you want to test this fixes your test from #13449 before I merge it?

**Comment by razajafri** [comment]
> @razajafri Do you want to test this fixes your test from #13449 before I merge it?

Ideally, I would like to do that but I won't be able to get to it until EOD today.

**Comment by revans2** [comment]
> > @razajafri Do you want to test this fixes your test from #13449 before I merge it?
> 
> Ideally, I would like to do that but I won't be able to get to it until EOD today.

I did a quick run and it is looking good. I would like to run all of our tests, but the ones that were failing before are passing with this change now.

**Pull request merged**